### PR TITLE
adVAE

### PR DESCRIPTION
**Pull request by TheMrGhostman**

### CODE REVIEW

File: scripts/experiments_images/adVAE.jl [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+        :extra_layers,
+        :gamma,
+        :lambda,
+        :mx,
+        :mz,
+        :lr, 
+        :decay,
+        :batch_size, 
+        :iters, 
+        :check_every,
```
Comment (vitskvara) [inline scripts/experiments_images/adVAE.jl]: I would not put theses parameters here as they never change and make the savename too long. You can just put them in the constructor or fit function directly.

File: scripts/experiments_images/adVAE.jl [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+function fit(data, parameters)
+	# define models (Generator, Discriminator)
+	advae = adVAE(;parameters...)
+
+	# define optimiser
+	try
+		global info, fit_t, _, _, _ = @timed fit!(advae |> gpu, data, parameters)
+	catch e
+		println("Error caught.")
```
Comment (vitskvara) [inline scripts/experiments_images/adVAE.jl]: Maybe print the error message?

File: scripts/experiments_images/adVAE.jl [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+        [30],
+        1:Int(1e8),
+    )
+
+    return NamedTuple{argnames}(map(x->sample(x,1)[1], par_vec))
+end
+
+function fit(data, parameters)
+	# define models (Generator, Discriminator)
+	advae = adVAE(;parameters...)
```
Comment (vitskvara) [inline scripts/experiments_images/adVAE.jl]: conv_advae?

File: scripts/experiments_images/adVAE.jl [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+		required = true
+		arg_type = String
+		help = "dataset"
+end
+parsed_args = parse_args(ARGS, s)
+@unpack dataset, max_seed = parsed_args
+
+modelname ="adVAE"
+
+function sample_params()
```
Comment (vitskvara) [inline scripts/experiments_images/adVAE.jl:32]: are these parameters for conv advae?

File: src/models/GANomaly.jl [hunk 5]
```diff
@@ -250,6 +250,8 @@ end
 computes generator and discriminator loss without additional pass through model
 """
 function validation_loss(g::Generator, d::Discriminator, real_input; weights=[1,50,1])
```
Comment (vitskvara) [inline src/models/GANomaly.jl]: I dont like the fact that switching to and from testmode is done here, it is very untransparent. Also, if the generator/discriminator is in testmode in the first place, after calling this function, it is not and the user can't possibly know it.

Comment (vitskvara) [reply src/models/GANomaly.jl]: I would prefer the switching to be done outside of the function.

File: src/models/GANomaly.jl [hunk 6]
```diff
@@ -271,18 +274,22 @@ end
 computes unscaled anomaly score A(x) = || E1(x) - E2(D(E1(x))) ||_1
 """
 function anomaly_score(generator::Generator, real_input; dims=3)
```
Comment (vitskvara) [inline src/models/GANomaly.jl]: The same as above, dont do testmode switching inside the function.

File: src/models/SkipGANomaly.jl [hunk 7]
```diff
@@ -185,6 +185,7 @@ end
 computes generator and discriminator loss without additional pass through model
 """
 function validation_loss(SkipGAN::SkipGANomaly, real_input; weights=[1,50,1])
+	testmode!(SkipGAN)
```
Comment (vitskvara) [inline src/models/SkipGANomaly.jl]: Again

Comment (vitskvara) [reply src/models/SkipGANomaly.jl]: And for all the other functions the same.

File: src/models/adVAE.jl [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+	# training info logger
+	history = MVHistory()
+	# prepare for early stopping
+	best_adVAE = deepcopy(advae)
+	patience = params.patience
+	best_val_loss = 1e10
+	val_batches = length(val_loader)
+
+	# ADAMW(η = 0.001, β = (0.9, 0.999), decay = 0) = Optimiser(ADAM(η, β), WeightDecay(decay))
+	opt = haskey(params, :decay) ? ADAMW(params.lr, (0.9, 0.999), params.decay) : ADAM(params.lr)
```
Comment (vitskvara) [inline src/models/adVAE.jl]: You need two optimisers, each for the separate set of parameters.

Comment (vitskvara) [reply src/models/adVAE.jl]: See the implementation of AAE.

File: src/models/adVAE.jl [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+	ps_step2 = Flux.params(advae.encoder)
+
+
+	progress = Progress(length(train_loader))
+	for (iter, X) in enumerate(train_loader)
+		# training step 1
+		loss_1, back = Flux.pullback(ps_step1) do
+            loss(advae|>gpu , getobs(X)|>gpu, γ=params.gamma, λ=params.lambda, mx=params.mx, mz=params.mz)
+		end
+		grad_step1 = back((1f0, 0f0, 0f0, 0f0))
```
Comment (vitskvara) [inline src/models/adVAE.jl:143]: Why do you do this?

File: src/models/adVAE.jl [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+
+	ps_step1 = Flux.params(advae.generator, advae.transformer)
+	ps_step2 = Flux.params(advae.encoder)
+
+
+	progress = Progress(length(train_loader))
+	for (iter, X) in enumerate(train_loader)
+		# training step 1
+		loss_1, back = Flux.pullback(ps_step1) do
+            loss(advae|>gpu , getobs(X)|>gpu, γ=params.gamma, λ=params.lambda, mx=params.mx, mz=params.mz)
```
Comment (vitskvara) [inline src/models/adVAE.jl]: What if you train on tabular data? Than this does not work. Move the model to gpu outside the fit function. Also, I would create a separate loss for tabular and image data, where the other is going to move them on gpu, and set it as an argument of the fit! function. See the vae.jl run scripts for tabular and image data.

File: src/models/adVAE.jl [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+					global iters = iter - params.check_every*params.patience
+					break
+				end
+			end
+		end
+		if iter == params.iters
+			global iters = params.iters
+		end
+	end
+	return history, best_adVAE, sum(map(p->length(p), ps_step1))+sum(map(p->length(p), ps_step2)), iters
```
Comment (vitskvara) [inline src/models/adVAE.jl]: If you have defined the model correctly with functors etc, you should only need to call `sum(map(p->length(p), Flux.params(best_adVAE)))`

File: src/models/adVAE.jl [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+function (advae::adVAE)(x)
+    μ, Σ = advae.encoder(x)
+    z = μ + Σ * randn(Float32)
+    xᵣ = advae.generator(z)
+end
+
+kl_divergence(μ, Σ) = 0.5 * sum(1 .+ log.(Σ.^2) - μ.^2  - Σ.^2) # actually it is -kl_divergence
+kl_divergence(μ₁, Σ₁, μ₂, Σ₂) = sum(log.(Σ₂ ./ Σ₁) + (Σ₁.^2 + (μ₁ - μ₂).^2) ./ (2*Σ₂.^2) .- 0.5)
+
+function loss(advae::adVAE, x; γ=1e-3, λ=1e-2, mx=1, mz=1)
```
Comment (vitskvara) [inline src/models/adVAE.jl:69]: Maybe, if this does not work, define separate loss functions for 1st and 2nd optimization step?

File: src/models/adVAE.jl [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+    μₜᵣ, Σₜᵣ = advae.encoder(xₜᵣ)
+
+    # 𝓛 for generator => 𝓛 = 𝓛_z + 𝓛_zₜ 
+    𝓛_z = Flux.Losses.mse(x, xᵣ) .+ γ * kl_divergence(μᵣ, Σᵣ)
+    𝓛_zₜ = max(0, (mx - Flux.Losses.mse(xᵣ, xₜᵣ))) + γ * max(0, (mz - kl_divergence(μₜᵣ, Σₜᵣ)))
+    𝓛 = 𝓛_z + 𝓛_zₜ
+    # 𝓛ₜ for transformer 
+    𝓛ₜ = kl_divergence(μ, Σ, μₜ, Σₜ)
+    # 𝓛ₑ for encoder
+    𝓛ₑ = Flux.Losses.mse(x, xᵣ) .+ γ * kl_divergence(μ, Σ)
```
Comment (vitskvara) [inline src/models/adVAE.jl]: Are you sure that the +- signs are all correct and every loss is written in such for that it is minimized? Because here it seems that the kl_divergence is actually maximized when it is with a - sign.

**Review by vitskvara — CHANGES_REQUESTED** [review]

File: scripts/experiments_images/adVAE.jl (hunk 3 above)
Comment (TheMrGhostman) [reply scripts/experiments_images/adVAE.jl]: yes, my bad

File: src/models/GANomaly.jl (hunk 5 above)
Comment (TheMrGhostman) [reply src/models/GANomaly.jl]: ok, i will change it

File: src/models/adVAE.jl (hunk 9 above)
Comment (TheMrGhostman) [reply src/models/adVAE.jl:143]: you mean this `grad_step1 = back((1f0, 0f0, 0f0, 0f0))` ?

it helps me trace individual losses. I returns all those losses but i am backpropagating only first one.

File: src/models/adVAE.jl (hunk 10 above)
Comment (TheMrGhostman) [reply src/models/adVAE.jl]: it works on tabular data. There is no problem. 
Moreover when you dont have gpu on your system, it automaticly works only on cpu with no error or warning.
But if you want i will chagne it.

File: src/models/adVAE.jl (hunk 11 above)
Comment (TheMrGhostman) [reply src/models/adVAE.jl]: changed

File: src/models/adVAE.jl (hunk 12 above)
Comment (TheMrGhostman) [reply src/models/adVAE.jl:69]: i will try

File: src/models/adVAE.jl (hunk 13 above)
Comment (TheMrGhostman) [reply src/models/adVAE.jl]: yes, in original paper there is mistake in +/- but i corrected it. mybe i should call `𝓛_kld instead of kl_divergence` because correctrly it is `-kld` 
-> `kl_divergence(μ, Σ) = 0.5 * sum(1 .+ log.(Σ.^2) - μ.^2  - Σ.^2) # actually it is -kl_divergence`

File: src/models/adVAE.jl (hunk 8 above)
Comment (TheMrGhostman) [reply src/models/adVAE.jl:123]: I think that Tomáš told me that when you have same hyperparameters (like lr and betas) you can use only one optimizer for both models.

**Review by vitskvara — APPROVED** [review]

**Comment by janfrancu** [comment]
Please bump pkg version to `0.3.0`.
